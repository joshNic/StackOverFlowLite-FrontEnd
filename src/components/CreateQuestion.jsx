/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable no-shadow */
import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { postQuestion } from "../actions/questionActions";

class CreateQuestion extends Component {
  state = {
    title: "",
    body: ""
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit = (e) => {
    e.preventDefault();
    const { title, body } = this.state;
    const { postQuestion } = this.props;
    const data = {
      question_title: title,
      question_body: body
    };
    postQuestion(data);
  }

  render() {
    const { title, body } = this.state;
    return (
      <div>
        <h1>Create Question</h1>
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputTitle"
            >
              Question Title
            </label>
            <input
              className="form-control form-control-lg"
              type="text"
              placeholder="Enter Question Title"
              id="inputTitle"
              name="title"
              onChange={this.onChange}
              value={title}
            />

            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputBody"
            >
              Question Description
            </label>
            <textarea
              className="form-control form-control-lg"
              rows="15"
              type="password"
              placeholder="Enter Question Description"
              id="inputBody"
              name="body"
              onChange={this.onChange}
              value={body}
            />

          </div>
          <div className="text-center">
            <button type="submit" className="btn btn-primary">
              Create Question
            </button>
          </div>
        </form>
      </div>
    );
  }
}

CreateQuestion.propTypes = {
  postQuestion: PropTypes.func.isRequired
};

export const mapStateToProps = state => ({
  question: state.questions.question
});

export const mapDispatchToProps = dispatch => ({
  postQuestion: data => dispatch(postQuestion(data))
});
export { CreateQuestion };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateQuestion);
