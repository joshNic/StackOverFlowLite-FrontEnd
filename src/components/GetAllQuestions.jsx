/* eslint-disable react/forbid-prop-types */
/* eslint-disable no-unused-expressions */
/* eslint-disable no-shadow */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import { getAll } from "../actions/questionActions";

class GetAllQuestions extends Component {
    
  componentWillMount(){
    const { getAll } = this.props;
    getAll();
  }

  questionValues = questions =>{
    let values;
    questions
      ? (
        values = questions.map(question => (
          <div key={question.question_id}>
            <div className="card border-primary mb-3">
              <div className="card-body">
                <Link to={{pathname: `/question/${question.question_id}`}}>
                  <h4 className="card-title">{question.question_title}</h4>
                </Link>
                <p className="card-text">{question.question_body}</p>
              </div>
            </div>
          </div>
        ))
      ) : null;
    
    return values;
  };
  
  render() {
    
    const { questions, loading } = this.props;
    return (
      <div>
        <h2>Questions</h2>
        {loading === true ? (
          <div className="text-center">
            <img
              src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
              className="loading__image"
              alt=""
            />
          </div>
        ) : 
          this.questionValues(questions)
        }  
      </div>
    );
  }
}

GetAllQuestions.propTypes = {
  questions: PropTypes.array.isRequired,
  getAll: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  questions: state.questions.questions,
  loading: state.questions.loading
});

export const mapDispatchToProps = dispatch => ({
  getAll: () => dispatch(getAll())
});
  
export { GetAllQuestions };
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GetAllQuestions);
