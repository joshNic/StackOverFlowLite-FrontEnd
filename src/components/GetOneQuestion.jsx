/* eslint-disable no-shadow */
/* eslint-disable no-unused-expressions */
import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { getOne } from "../actions/questionActions";

class GetOneQuestion extends Component {
    
  componentWillMount(){
    const { match, getOne } = this.props;
    const { id } = match.params;
    getOne(id);
  }

  questionValues = question =>{
    let values;
    question
      ? (
        values = (
          <div className="card border-primary mb-3">
            <div className="card-body">
              <h4 className="card-title">{question.question_title}</h4>
              <p className="card-text">{question.question_body}</p>
            </div>
          </div>
        )
      ) : null;
    
    return values;
  };
    
  render() {
    const { question, loading } = this.props;
    return (
      <div>
        <h2>Question</h2>
        {loading === true ? (
          <div className="text-center">
                  
            <img
              src="https://media.giphy.com/media/jAYUbVXgESSti/giphy.gif"
              className="loading__image"
              alt=""
            />
          </div>
        ) : 
          this.questionValues(question)
        }  
      </div>
    );
  }
}

GetOneQuestion.propTypes = {
  question: PropTypes.shape({}).isRequired,
  match: PropTypes.shape({}).isRequired,
  getOne: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  questions: state.questions.questions,
  loading: state.questions.loading,
  question: state.questions.question,
});
  
export const mapDispatchToProps = dispatch => ({
  getOne: (id) => dispatch(getOne(id))
});

export { GetOneQuestion };
  
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GetOneQuestion);