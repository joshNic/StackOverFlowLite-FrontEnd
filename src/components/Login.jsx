/* eslint-disable no-shadow */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { login } from "../actions/questionActions";

class Login extends Component {
  state = {
    email: "",
    password: ""
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit = (e) => {
    e.preventDefault();

    const { email, password } = this.state;
    const { login } = this.props;
    const data = {
      email: email,
      password: password
    };
    login(data);
  }
  
  render() {
    const { email, password } = this.state;

    return (
      <div>
        <h1>Login</h1>
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputEmail"
            >
              Email
            </label>
            <input
              className="form-control form-control-lg"
              type="email"
              placeholder="Enter Email"
              id="inputEmail"
              name="email"
              onChange={this.onChange}
              value={email}
            />
            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputPassword"
            >
              Password
            </label>
            <input
              className="form-control form-control-lg"
              type="password"
              placeholder="Enter password"
              id="inputPassword"
              name="password"
              onChange={this.onChange}
              value={password}
            />
          </div>
          <div className="text-center">
            <button type="submit" className="btn btn-primary">
              Login
            </button>
          </div>
        </form>
      </div>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired
};

export const mapDispatchToProps = dispatch => ({
  login: data => dispatch(login(data))
});
export { Login };

export default connect(
  null,
  mapDispatchToProps
)(Login);
