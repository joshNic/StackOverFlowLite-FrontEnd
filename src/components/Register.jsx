/* eslint-disable no-shadow */
/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { register } from "../actions/questionActions";

class Register extends Component {
  state = {
    email: "",
    password: ""
  };

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = e => {
    e.preventDefault();
    const { email, password } = this.state;
    const { register } = this.props;
    const data = {
      user_email: email,
      user_password: password
    };
    register(data);
  };

  render() {
    const { email, password } = this.state;

    return (
      <div>
        <h1> Create Account</h1>
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputEmail"
            >
              Email
            </label>
            <input
              className="form-control form-control-lg"
              type="email"
              placeholder="Enter Email"
              id="inputEmail"
              name="email"
              onChange={this.onChange}
              value={email}
            />
            <label
              className="col-form-label col-form-label-lg"
              htmlFor="inputPassword"
            >
              Password
            </label>
            <input
              className="form-control form-control-lg"
              type="password"
              placeholder="Enter password"
              id="inputPassword"
              name="password"
              onChange={this.onChange}
              value={password}
            />
          </div>
          <div className="text-center">
            <button type="submit" className="btn btn-primary">
              Create
            </button>
          </div>
        </form>
      </div>
    );
  }
}

Register.propTypes = {
  register: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  question: state
});

export const mapDispatchToProps = dispatch => ({
  register: data => dispatch(register(data))
});

export { Register };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Register);
