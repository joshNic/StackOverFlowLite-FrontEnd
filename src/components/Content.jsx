import React from "react";
import { Route } from "react-router-dom";

import GetAllQuestions from "./GetAllQuestions";
import CreateQuestion from "./CreateQuestion";
import GetOneQuestion from "./GetOneQuestion";
import Login from "./Login";
import Register from "./Register";

const Content = () => 
  (
    <div>
      <div className="jumbotron">
        <Route exact path="/login" component={Login} />
        <Route exact path="/register" component={Register} />
        <Route exact path="/create" component={CreateQuestion} /> 
        <Route exact path="/questions" component={GetAllQuestions} />
        <Route exact path="/" component={GetAllQuestions} />
        <Route path="/question/:id" component={GetOneQuestion} />
      </div>
    </div>
  );

export default Content;

