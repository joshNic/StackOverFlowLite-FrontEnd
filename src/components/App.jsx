import React from "react";

import "../App.css";
import Footer from "../commons/Footer";
import Header from "../commons/Header";
import Content from "./Content";


const  App = () => 
  (
    <div>
      <div className="container">
        
        <header>
          <Header /> 
        </header>
        <Content />
      </div>
      <footer className="container">
        <Footer />
      </footer>
    </div>
  );

export default App;
