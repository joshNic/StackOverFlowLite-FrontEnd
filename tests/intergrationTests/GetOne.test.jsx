import "babel-polyfill";
import React from "react";
import { shallow } from "enzyme";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";


import { GetOneQuestion, mapDispatchToProps } from "../../src/components/GetOneQuestion";

describe("<Content />", () => {

  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const initialUserState = {
    questions:[],
    question:{},
    loading:true,
    response:{}   
  };

  const store = mockStore({ ...initialUserState });
  const initialState = {
    title: "",
    body: "",
  };
 
  const editor = shallow(<GetOneQuestion loading={false} questions={[initialState]} getOne={jest.fn()} store={store} match={{ params: { id: 2 } }} />);
  
  it("should render without crashing", () => {
    
    expect(editor).toMatchSnapshot();
  });

  test("it triggers the postQuestion action", () => {
    const mock = jest.fn();
    const mapper = mapDispatchToProps(mock);
    mapper.getOne(2);
    expect(mock).toHaveBeenCalled();
  });
});
