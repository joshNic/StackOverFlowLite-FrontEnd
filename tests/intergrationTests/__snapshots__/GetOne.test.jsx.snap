// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Content /> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GetOneQuestion
    getOne={
      [MockFunction] {
        "calls": Array [
          Array [
            2,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    loading={false}
    match={
      Object {
        "params": Object {
          "id": 2,
        },
      }
    }
    questions={
      Array [
        Object {
          "body": "",
          "title": "",
        },
      ]
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Question
        </h2>,
        undefined,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Question",
        },
        "ref": null,
        "rendered": "Question",
        "type": "h2",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Question
          </h2>,
          undefined,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Question",
          },
          "ref": null,
          "rendered": "Question",
          "type": "h2",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
